### ISSUE TITLE
Exception throwing/catching significantly slower on Linux than on Windows

### ISSUE THREAD

**Issue by stephentoub** [dotnet/coreclr#1622]
As a baseline, this program takes approximately the same amount of time per iteration on my Linux VM  as it does on the host Windows machine:
```C#
using System;
using System.Diagnostics;
using System.Runtime.CompilerServices;

public static class Test
{
    public static void Main()
    {
        var sw = new Stopwatch();
        while (true)
        {
            sw.Restart();
            for (int i = 0; i < 100000000; i++)
            {
                Nop();
            }
            Console.WriteLine(sw.Elapsed);
        }
    }

    [MethodImpl(MethodImplOptions.NoInlining)]
    private static void Nop() { }
}
```
In contrast, this program that throws/catches an exception per iteration:
```C#
using System;
using System.Diagnostics;

public static class Test
{
    public static void Main()
    {
        var sw = new Stopwatch();
        Exception exc = new Exception();
        while (true)
        {
            sw.Restart();
            for (int i = 0; i < 1000; i++)
            {
                try { throw exc; }
                catch { }
            }
            Console.WriteLine(sw.Elapsed);
        }
    }
}
```
takes ~1.3 seconds on my Linux VM vs ~0.008 on my Windows machine, making it ~160x slower.

**Comment by stephentoub** [comment]
cc: @janvorli, @kouvel, @jkotas

**Comment by janvorli** [comment]
I have tried to profile it using the oprofile Linux profiler. Interestigly, 87% of the time is spent in the Linux kernel. And I can see that the corerun eats 100% of the CPU when running the test, so it is not just idle waiting.
We'll need to profile it on a physical machine to get details on what syscalls are actually being called. The oprofile cannot do that on a VM.

**Comment by janvorli** [comment]
I have looked at the profiler generated data once again and I have noticed that there is quite high number of samples (although only 1.2 percent of the total) in  _IO_vfscanf. Setting breakpoint on this function revealed that it is called from pthread_getattr_np, which in turn is called from PAL_GetStackBase / PAL_GetStackLimit. And that one is called very often.
So I have experimentally created a thread local caching variable for the returns of these functions so that we call the pthread_getattr_np only once per thread lifetime to see if it makes any difference.
And it really did a huge difference. On my VM, the time reported by the testing app from @stephentoub  dropped from 2.16 secs to 0.46 secs.

**Comment by janvorli** [comment]
I've forgotten to mention that I have made these measurements in debug build.

**Comment by janvorli** [comment]
@stephentoub I have measured the code with my change in release build on the same VM. The time reported is 0.03 secs. That gets us to approx 4x slower on Linux.
I will prepare a PR for the change soon.

**Comment by stephentoub** [comment]
Great improvement, thanks!

I assume we'll leave this issue open in order to track down the remaining difference? Or do you believe this is the best we can do?

**Comment by janvorli** [comment]
@stephentoub I have drilled into it a bit more. The top 10 time (summing to 63%) is divided this way:
```
34.5% in C++ exception unwinding in libgcc_s.so.1
 7.0% in libunwind8
 6.3% in libstdc++
 4.1% in kernel
 2.5% in __memcpy_sse2_unaligned
 2.4% in pthread_mutex_unlock
 1.8% in __tls_get_addr
 1.7% in UnicodeToUTF8
 1.5% in pthread_mutex_lock
 1.2% in memset
```
The remaining 37% is divided between a lot of small contributions.
It is not clear what is consuming time in the libstdc++, I've tried to install debug symbols, but the OProfile doesn't show any and shows the whole library as single entry.

Given this list, it seems it is unlikely to get significantly better. We can do nothing about the C++ unwinding and libunwind which are the major contributors.
